go/vt/topo/metadata.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"regexp"
+	"strings"
+
+	"vitess.io/vitess/go/event"
+	"vitess.io/vitess/go/vt/topo/events"
+)
+
+// UpsertMetadata sets the key/value in the metadata if it doesn't exist, otherwise it updates the content
+func (ts *Server) UpsertMetadata(ctx context.Context, key string, val string) error {
+	keyPath := path.Join(MetadataPath, key)
+
+	_, _, err := ts.globalCell.Get(ctx, keyPath)
+	status := "updated"
+
+	if err != nil {
+		if !IsErrType(err, NoNode) {
+			return err
+		}
+
+		status = "created"
+	}
+
+	// nil version means that it will insert if keyPath does not exist
+	if _, err := ts.globalCell.Update(ctx, keyPath, []byte(val), nil); err != nil {
+		return err
+	}
+
+	dispatchEvent(keyPath, status)
+	return nil
+}
+
+// GetMetadata retrieves all metadata value that matches the given key regular expression. If empty all values are returned.
+func (ts *Server) GetMetadata(ctx context.Context, keyFilter string) (map[string]string, error) {
+	keys, err := ts.globalCell.ListDir(ctx, MetadataPath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	keyPattern := strings.Replace(regexp.QuoteMeta(keyFilter), "%", ".*", -1)
+	keyPattern = fmt.Sprintf("^%s$", keyPattern)
+	re := regexp.MustCompile(keyPattern) // Can never fail
```

Comment:
Are we 100% sure this is a compatible conversion from the LIKE syntax to regexp? Itt's fine for this use case but if there's a more robust and generic way to do this, we might get some mileage out of making it a common function in the Vitess code base.

Reply:
I found this link with examples of the LIKE syntax: https://www.w3schools.com/sql/sql_like.asp

I we also need to support `_`.

I agree that extracting a "LIKE to regexp" converter into a separate file and a more shared place and then writing specific unit tests for that would be nice. I'm not sure we need to make that 100% compliant in this PR though. We can iterate on that once we have it in a central place with a nice suite of unit tests.

Reply:
Yep -- totally agree 100%

Reply:
Extracted functionality to separate file with unit tests. Also supporting '_' character and escaping using "\"